component/s3/file/s3.go

```diff
@@ -0,0 +1,213 @@
+package s3
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/grafana/agent/pkg/flow/rivertypes"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	aws_config "github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/grafana/agent/component"
+	"golang.org/x/net/context"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name:    "s3.file",
+		Args:    Arguments{},
+		Exports: Exports{},
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+// S3 handles reading content from a file located in S3
+type S3 struct {
+	mut        sync.Mutex
+	opts       component.Options
+	args       Arguments
+	health     *component.Health
+	watcher    *watcher
+	updateChan chan []byte
+	errorChan  chan error
+	content    string
+	cancel     context.CancelFunc
+}
+
+var (
+	_ component.Component       = (*S3)(nil)
+	_ component.HealthComponent = (*S3)(nil)
+)
+
+// New initializes the s3 component
+func New(o component.Options, args Arguments) (*S3, error) {
+	s3cfg, err := generateS3Config(args)
+	if err != nil {
+		return nil, err
+	}
+
+	s3Client := s3.NewFromConfig(*s3cfg, func(s3o *s3.Options) {
+		s3o.UsePathStyle = args.Options.UsePathStyle
+	})
+
+	bucket, file := getPathBucketAndFile(args.Path)
+	s := &S3{
+		opts:       o,
+		args:       args,
+		health:     &component.Health{},
+		updateChan: make(chan []byte),
+		errorChan:  make(chan error),
+	}
```

Comment:
Can this try to do an initial load at construction time so it immediately fails if the file can't be retrieved? (local.file does this)

Reply:
I didn't do it that way since we don't have a parent context, especially given it's going over the network and could take some time.

Reply:
I think that's probably fine to start with, but it does mean other components will start with referencing an empty string until the file gets retrieved.

Reply:
Changed it up so this downloads synchronously on New.